provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+import { EventEmitter } from 'events';
+import { errors, X509, SharedAccessSignature } from 'azure-iot-common';
+import * as machina from 'machina';
+import * as Provisioning from './transport_interface';
+import * as dbg from 'debug';
+const debug = dbg('azure-device-provisioning:transport-fsm');
+
+export interface TransportHandlers {
+  _doConnectForFsm(callback: (err?: Error) => void): void;
+  _doDisconnectForFsm(callback: (err?: Error) => void): void;
+  _doFirstRegistrationRequestForFsm(registrationId: string, authorization: SharedAccessSignature | X509 | string, requestBody: any, forceRegistration: boolean, callback: (err?: Error, responseBody?: any, result?: any, pollingInterval?: number) => void): void;
+  _doOperationStatusQueryForFsm(registrationId: string, operationId: string, callback: (err?: Error, responseBody?: any, result?: any, pollingInterval?: number) => void): void;
+  _getErrorFromResultForFsm(result: any): any;
+}
+
+export class TransportStateMachine extends EventEmitter implements Provisioning.Transport, TransportHandlers{
+  protected _apiVersion: string = '2017-08-31-preview';
+  private _fsm: machina.Fsm;
+  private _pollingTimer: any;
+  private _registrationCallback: Provisioning.ResponseCallback;
+
+  /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_001: [ The `constructor` shall accept no arguments ] */
+  constructor() {
+    super();
+
+    this._fsm = new machina.Fsm({
+      namespace: 'Provisioning',
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (callback, err) => {
+            debug('entering disconnected state');
+            this._pollingTimer = null;
+            this._registrationCallback = null;
+            if (callback) {
+              callback(err);
+            }
+          },
+          connect: (callback) => {
+            this._fsm.transition('connecting', callback);
+          },
+          register: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_010: [ `register` shall `connect` the transport if it is not connected. ] */
+            this._fsm.handle('connect', (err) => {
+              if (err) {
+                this._fsm.transition('disconnecting', callback, err);
```

Comment:
>this._fsm.transition('disconnecting', callback, err); [](start = 16, length = 53)

connect itself should handle the transitions. these should only handle the call to the callback

Reply:
yes, and this only points to the need for a more comprehensive state transition review.

---
In reply to: [146400668](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r146400668) [](ancestors = 146400668)